### PR TITLE
[WIP] Standard symfony configuration management for the datagrid and datasource

### PR DESCRIPTION
**Pull request by Befooz**
I would like to suggest another way to management datagrid and datasource configurations.
Configurations of the bundles will collect once (currently library find and parse it for each request) and store in the container.
Also we have the ability to use resource caching, validate them with the Configuration and also configure them at the application layer (without bundle).

What do you think? If you accept the changes, I can prepare the code for datasource and modify tests.

### CODE REVIEW

**Comment by Befooz** [comment]
ping @szymach @JarekW

**Comment by Befooz** [comment]
ping @chives and @rn0

**Comment by chives** [comment]
@Befooz you realize that this is a huge BC break, right? We could introduce your solution as an additional one, but not as a replacement for current configuration loading. Datagrid configurations read from symfony configuration may have higher priority than old style configurations, if the same datagrid name appears in both places.

**Comment by Cosmologist** [comment]
@chives, the solution is fully backward-compatible.
Configuration files are stored in the same places and have the same structure.
It was changed only the loading and processing workflow.

**Comment by Befooz** [comment]
@chives, the solution is fully backward-compatible.
Configuration files are stored in the same places and have the same structure.
It was changed only the loading and processing workflow.

**Comment by chives** [comment]
@Befooz you're right! I haven't noticed that at first glance

File: DataGrid/Extension/Configuration/EventSubscriber/ConfigurationBuilder.php [hunk 1]
```diff
@@ -19,84 +19,53 @@
 class ConfigurationBuilder implements EventSubscriberInterface
 {
     /**
-     * @var \Symfony\Component\HttpKernel\KernelInterface
+     * Datagrid configuration
+     *
+     * @var array
      */
-    protected $kernel;
+    protected $datagridConfiguration;
```
Comment (chives) [inline DataGrid/Extension/Configuration/EventSubscriber/ConfigurationBuilder.php]: i think we should name this variable `datagridsConfigurations` as long as it contains all datagrids' configurations

**Comment by Befooz** [comment]
@chives, You will be ready to accept changes?
Then I'll be ready to spend time to modifing datasource and tests.

File: DataGrid/Extension/Configuration/EventSubscriber/ConfigurationBuilder.php (hunk 1 above)
Comment (Befooz) [inline DataGrid/Extension/Configuration/EventSubscriber/ConfigurationBuilder.php]: It's right!

**Comment by fduch** [comment]
@chives there is 2 main point of this patch:
1. Reduce configuration load time by using caching mechanisms based on Symfony configuration component
2. Provide ability to override datagrid-configs from bundles on application level. This one seems to be helpful for testing purposes or decoupling configurations through the different environments. Btw it also looks a little bit more Symfony-stylish)

**Comment by chives** [comment]
@Befooz yes, i'm definitely for this change. Also https://github.com/fsi-open/datasource-bundle could benefit from similar improvements. We have only one side effect of this change. If bundle A defines some datagrid and bundle B defines datagrid with the same name, then these 2 configurations will be unintentionally merged in effective configuration. I'm not sure if we can do anything about that, but we definitely should write about it in the docs. This could be a feature allowing to extend another datagrid's configuration by adding or changing some columns, but it cannot be used to remove any previously defined column.

**Comment by fduch** [comment]
@Befooz @chives i guess it can be dangerous to allow 2 bundles to have grid configurations with the same name merged into one. 
In case of simple collision such configuration merge can lead to unexpected grid structure (consider 2 independent bundles with absolutely different datagrids with name 'client'). May be we should throw exception when such name collision is detected in `prepend` method? The bundles containing these configurations will be incompatible (when datagrid-bundle is activated) but we explicitly prevent inconsistent merge result. Obviously we'll should note this case in docs)

**Comment by szymach** [comment]
@Befooz Hello, we are cleaning up our open source project and wanted to ask if there is any possibility that you will finish this PR? Otherwise we will close this PR.

**Comment by szymach** [comment]
Closed due to inactivity

**Pull request closed without merging**